### PR TITLE
Make it possible to use AVD

### PR DESCRIPTION
**Pull request by hardening**
This PR reworks ARM so that we can connect directly with a bearer token.

### CODE REVIEW

**Comment by hardening** [comment]
@freerdp-bot can you test correctly please ?

File: libfreerdp/core/gateway/arm.c [hunk 1]
```diff
@@ -311,6 +317,368 @@ static char* arm_create_request_json(rdpArm* arm)
 	return message;
 }
 
+static WINPR_CIPHER_CTX* treatAuthBlob(const BYTE* pbInput, size_t cbInput)
+{
+	WINPR_CIPHER_CTX* ret = NULL;
+	char algoName[100];
+
+	SSIZE_T algoSz = ConvertWCharNToUtf8((WCHAR*)pbInput, cbInput / 2, algoName, sizeof(algoName));
```
Comment (akallabeth) [inline libfreerdp/core/gateway/arm.c:325]: `cbInput / sizeof(WCHAR)`

**Review by akallabeth — CHANGES_REQUESTED** [review]
some stuff found, need to do a thorough review later.
as for the parsing of redirection certificate, is that not already available in `redirection.c`?

**Comment by hardening** [comment]
Hum correct the certificate has the same format in the redirection, I may expose and reuse these functions (changing them a little to suit my needs)

**Comment by hardening** [comment]
@akallabeth fixed your remarks, and mutualized the code with redirection where it was possible

File: libfreerdp/core/redirection.c [hunk 2]
```diff
@@ -384,15 +385,15 @@ static BOOL rdp_redirection_read_target_cert(rdpRedirection* redirection, const
 				}
 				break;
 			default: /* ignore unknown fields */
-				WLog_WARN(TAG,
-				          "Unknown TargetCertificate field type %" PRIu32 ", encoding %" PRIu32
-				          " of length %" PRIu32,
-				          type, encoding, plength);
+				WLog_VRB(TAG,
```
Comment (akallabeth) [inline libfreerdp/core/redirection.c:388]: `trace` might be a bit too hidden, maybe `debug` instead?

File: libfreerdp/crypto/base64.c [hunk 3]
```diff
@@ -26,17 +26,26 @@
...
+static char* base64_encode_ex(const char* alphabet, const BYTE* data, size_t length, BOOL pad,
+                              BOOL crLf, size_t lineSize)
 {
 	int c;
 	const BYTE* q;
 	char* p;
 	char* ret;
 	int i = 0;
 	int blocks;
+	int outLen = (length + 3) * 4 / 3;
```
Comment (akallabeth) [inline libfreerdp/crypto/base64.c:38]: unchecked truncation from `size_t` to `int`

File: libfreerdp/crypto/base64.c [hunk 4]
```diff
@@ -26,17 +26,26 @@
...
+	int extra = 0;
+	if (crLf)
+	{
+		int nCrLf = (outLen + lineSize - 1) / lineSize;
```
Comment (akallabeth) [inline libfreerdp/crypto/base64.c:42]: unchecked truncation from `size_t` to `int`

File: libfreerdp/crypto/base64.c [hunk 5]
```diff
@@ -26,17 +26,26 @@
...
+		extra = nCrLf * 2;
+	}
+	int outCounter = 0;
 
 	q = data;
-	p = ret = (char*)malloc((length + 3) * 4 / 3 + 1);
+	p = ret = (char*)malloc(outLen + extra + 1);
```
Comment (akallabeth) [inline libfreerdp/crypto/base64.c:48]: better use `1ull` to enforce `64bit` size

File: libfreerdp/core/gateway/arm.c [hunk 6]
```diff
@@ -311,6 +318,319 @@ static char* arm_create_request_json(rdpArm* arm)
 	return message;
 }
 
+static WINPR_CIPHER_CTX* treatAuthBlob(const BYTE* pbInput, size_t cbInput)
+{
+	WINPR_CIPHER_CTX* ret = NULL;
+	char algoName[100];
```
Comment (akallabeth) [inline libfreerdp/core/gateway/arm.c:324]: maybe initialize with `={0};`

File: libfreerdp/core/gateway/arm.c [hunk 7]
```diff
@@ -311,6 +318,319 @@ static char* arm_create_request_json(rdpArm* arm)
...
+	algoName[algoSz] = 0;
+	if (strcmp(algoName, "AES") != 0)
+	{
+		WLog_ERR(TAG, "only AES is supported for now");
+		return NULL;
+	}
+
+	cbInput -= (algoSz + 1) * sizeof(WCHAR);
+	/* BCRYPT_KEY_DATA_BLOB_HEADER */
+	PBCRYPT_KEY_DATA_BLOB_HEADER header =
```
Comment (akallabeth) [inline libfreerdp/core/gateway/arm.c:343]: Maybe a good idea to wrap the `pbInput` into a static stream?

File: libfreerdp/core/gateway/arm.c [hunk 8]
```diff
@@ -311,6 +318,319 @@ static char* arm_create_request_json(rdpArm* arm)
...
+	*ppOut = NULL;
+	*pcbOut = 0;
+
+	/* encode to base64 with crlf */
+	char* b64encoded = crypto_base64_encode_ex(pin, cbIn, TRUE);
+	if (!b64encoded)
+		return FALSE;
+
+	/* and then convert to Unicode */
+	size_t outSz;
```
Comment (akallabeth) [inline libfreerdp/core/gateway/arm.c:416]: better initialize the variable

File: libfreerdp/core/gateway/arm.c [hunk 9]
```diff
@@ -311,6 +318,319 @@ static char* arm_create_request_json(rdpArm* arm)
...
+	}
+
+	if (!cJSON_IsArray(interface))
+	{
+		WLog_ERR(TAG, "expecting interface to be an Array");
+		goto out;
+	}
+
+	int interfaceSz = cJSON_GetArraySize(interface);
+	if (!interfaceSz)
```
Comment (akallabeth) [inline libfreerdp/core/gateway/arm.c:585]: `interfaceSz` is an integer, better check `interfaceSz == 0` - or `<= 0`?)

**Review by akallabeth — CHANGES_REQUESTED** [review]
overall LGTM

File: libfreerdp/core/gateway/arm.c [hunk 10]
```diff
@@ -311,6 +318,319 @@ static char* arm_create_request_json(rdpArm* arm)
...
+{
+	BOOL ret = FALSE;
+	BYTE* output = NULL;
+	BYTE* finalOutput = NULL;
+
+	/* let's prepare the encrypted password, first we do a
+	 *    cipheredPass = AES(redirectedAuthBlob, toUtf16(passwd))
+	 */
+	size_t wpasswdLen = 0;
+	WCHAR* wpasswd = ConvertUtf8ToWCharAlloc(settings->Password, &wpasswdLen);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/arm.c:438]: better use the `freerdp_settings_get_string_*` functions directly

File: libfreerdp/core/redirection.c (hunk 2 above)
Comment (hardening) [reply libfreerdp/core/redirection.c:388]: Verbose is fine, we're not interested by most of the exported items (I'm even wondering if there should have a log).

File: libfreerdp/core/redirection.h [hunk 11]
```diff
@@ -31,12 +31,20 @@ typedef struct rdp_redirection rdpRedirection;
...
+#define CERT_ctl_file_element 34
+#define CERT_keyid_file_element 35
+
 FREERDP_LOCAL int rdp_redirection_apply_settings(rdpRdp* rdp);
 
 FREERDP_LOCAL state_run_t rdp_recv_enhanced_security_redirection_packet(rdpRdp* rdp, wStream* s);
 FREERDP_LOCAL BOOL
 rdp_write_enhanced_security_redirection_packet(wStream* s, const rdpRedirection* redirection);
 
+FREERDP_LOCAL BOOL rdp_redirection_read_target_cert(rdpCertificate** ptargetCertificate,
```
[the remark this answers is not in the document]
Comment (hardening) [reply libfreerdp/core/redirection.h:45]: The redirection certificate is always in DER format, so from my PoV it's not needed...

File: libfreerdp/core/gateway/arm.c (hunk 6 above)
Comment (hardening) [reply libfreerdp/core/gateway/arm.c:324]: No need to zero the full array, we do set the end of string later.

File: libfreerdp/core/gateway/arm.c (hunk 7 above)
Comment (hardening) [reply libfreerdp/core/gateway/arm.c:343]: Hum good point, I'm wondering if the exporting code really cares about the endianess...

File: libfreerdp/core/gateway/arm.c (hunk 8 above)
Comment (hardening) [reply libfreerdp/core/gateway/arm.c:439]: well it is set at the next line, so...

**Review by akallabeth — APPROVED** [review]
LGTM

**Pull request merged**